File: scripts/import-members.py
import pandas as pd
import requests
import json
from typing import Dict, List, Optional, Union, Any

def clean_string(value: Any) -> Optional[str]:
    """Clean up string values, return None if empty."""
    if pd.isna(value):
        return None
    return str(value).strip()

def clean_class_year(value: Any) -> Optional[int]:
    """Clean up class year, ensuring it's a valid integer."""
    if pd.isna(value):
        return None
    try:
        if isinstance(value, str):
            # Remove any non-numeric characters
            numeric = ''.join(filter(str.isdigit, value))
            return int(numeric)
        return int(value)
    except (ValueError, TypeError):
        return None

def map_membership_type(value: str) -> str:
    """Map membership type to valid enum values."""
    value = value.lower() if isinstance(value, str) else ""
    if "honorary" in value:
        return "HONORARY"
    if "student" in value:
        return "STUDENT"
    if "ordinary ii" in value:
        return "ORDINARY_II"
    return "ORDINARY_I"

def rename_columns(df: pd.DataFrame) -> pd.DataFrame:
    """Rename unnamed columns based on the first row."""
    # Get the first row which contains our column names
    column_names = df.iloc[0]
    
    # Create a mapping of current column names to desired column names
    column_mapping = {}
    for i, col in enumerate(df.columns):
        if pd.notna(column_names[i]):
            column_mapping[col] = str(column_names[i]).strip()
    
    # Rename the columns
    df = df.rename(columns=column_mapping)
    
    # Drop the first row as it was our column names
    df = df.iloc[1:].reset_index(drop=True)
    
    return df

def validate_row(row: Dict[str, Any]) -> tuple[bool, Optional[str]]:
    """Validate a single row of data."""
    # Map Excel columns to our required fields
    required_mappings = {
        "English Name": ["English Name", "Name Box"],
        "Email": ["Email"],
        "Class": ["Class"],
        "Faculty": ["Faculty"],
        "Major": ["Major"]
    }
    
    # Check each required field
    for field, possible_columns in required_mappings.items():
        found = False
        for col in possible_columns:
            if col in row and not pd.isna(row[col]):
                found = True
                break
        if not found:
            return False, f"Missing required field: {field}"
    
    # Validate class year
    class_value = None
    for col in required_mappings["Class"]:
        if col in row:
            class_value = clean_class_year(row[col])
            if class_value:
                break
    
    if not class_value:
        return False, "Invalid or missing Class year"
    
    return True, None

def transform_row(row: Dict[str, Any]) -> Dict[str, Any]:
    """Transform a row into the API format."""
    # Get English name from possible columns
    english_name = clean_string(row.get("English Name"))
    preferred_name = clean_string(row.get("Preferred Name"))
    
    # Use preferred name if available, otherwise use English name
    final_name = preferred_name if preferred_name else english_name
    
    # Get membership type and fee
    membership_type = map_membership_type(str(row.get("Membership", "ORDINARY_I")))
    membership_fee = 500 if membership_type in ["ORDINARY_I", "ORDINARY_II"] else 0
    
    # Parse transaction date if available
    transaction_date = None
    if not pd.isna(row.get("Date of transaction")):
        try:
            transaction_date = pd.to_datetime(row["Date of transaction"]).isoformat()
        except:
            pass
    
    # Parse registration date if available
    registration_date = None
    if not pd.isna(row.get("Date of registration")):
        try:
            registration_date = pd.to_datetime(row["Date of registration"]).isoformat()
        except:
            pass
    
    # Map the row to our API format
    return {
        # Personal Information
        "salutation": clean_string(row.get("Salutation", "Mr")),
        "englishName": final_name,
        "preferredName": preferred_name,
        "chineseName": clean_string(row.get("Chinese Name")),
        "gender": clean_string(row.get("Gender", "prefer_not_to_say")),
        
        # Academic Information
        "class": clean_class_year(row.get("Class")),
        "faculty": clean_string(row.get("Faculty")),
        "major": clean_string(row.get("Major")),
        "cusid": clean_string(row.get("CUSID", "LEGACY")),
        "studentIdImage": "LEGACY",  # Default value for imported members
        
        # Professional Information
        "employer": clean_string(row.get("Employer")),
        "position": clean_string(row.get("Position")),
        "industry": clean_string(row.get("Industry", "")),
        
        # Contact Information
        "email": clean_string(row.get("Email")),
        "phoneNumber": str(row.get("Phone Number", "")),
        "address": clean_string(row.get("Address")),
        
        # Membership Details
        "membershipType": membership_type,
        "membershipFee": membership_fee,
        "transactionDate": transaction_date,
        "dateOfRegistration": registration_date,
        
        # Status (all imported members are verified)
        "isVerified": True,
        "status": "APPROVED"
    }

def import_members():
    try:
        print("Reading Excel file...")
        # Read the Excel file, skipping empty rows
        df = pd.read_excel(
            "instructions/membersdata/Membership Book.xlsx",
            na_values=['', 'NA', 'N/A'],
            keep_default_na=True
        )
        
        # Rename columns using the first row
        df = rename_columns(df)
        
        # Print column names for debugging
        print("\nDetected columns:")
        print(df.columns.tolist())
        
        # Take only first 2 rows for testing
        test_df = df.head(2)
        print("\nProcessing first 2 rows:")
        print(test_df.to_string())
        
        # Convert rows to dictionaries and validate
        valid_rows = []
        invalid_rows = []
        
        for index, row in test_df.iterrows():
            row_dict = row.to_dict()
            print(f"\nValidating row {index + 1}:")
            print("Raw data:", json.dumps(row_dict, default=str))
            
            is_valid, error = validate_row(row_dict)
            if is_valid:
                print("✅ Row is valid")
                transformed = transform_row(row_dict)
                print("Transformed data:", json.dumps(transformed, indent=2))
                valid_rows.append(transformed)
            else:
                print(f"❌ Row has validation error: {error}")
                invalid_rows.append({
                    "row": index + 2,  # +2 because Excel is 1-based and we have a header row
                    "error": error
                })
        
        # Log validation errors if any
        if invalid_rows:
            print("\nValidation errors found:")
            for row_error in invalid_rows:
                print(f"Row {row_error['row']}: {row_error['error']}")
            
            if not valid_rows:
                raise ValueError("No valid rows found in Excel file")
            print("\nProceeding with valid rows only...\n")
        
        # Ask for confirmation before importing
        print("\nReady to import the following members:")
        for member in valid_rows:
            print(f"\n{member['englishName']} ({member['email']}):")
            print(json.dumps(member, indent=2))
        
        confirm = input("\nProceed with import? (y/n): ")
        if confirm.lower() != 'y':
            print("Import cancelled")
            return
        
        print("\nImporting members to database...")
        # Call the Next.js API endpoint directly
        response = requests.post(
            "http://localhost:3000/api/import-members",  # We'll create this endpoint
            json=valid_rows,  # Send the array directly
            headers={
                "Content-Type": "application/json",
                "x-api-key": "your-secret-key"  # We'll add this for security
            }
        )
        
        if not response.ok:
            print("\nAPI Response:", response.text)
            raise ValueError(f"API request failed: {response.text}")
        
        try:
            result = response.json()
            print("\nImport completed:")
            print(f"Total processed: {result['totalProcessed']}")
            print(f"Successful: {result['successful']}")
            print(f"Failed: {result['failed']}")
            
            # Log detailed results
            print("\nDetailed results:")
            for r in result["results"]:
                if r["success"]:
                    print(f"✅ {r['email']}: Successfully imported")
                    print(f"  Member ID: {r['memberId']}")
                    print(f"  Dummy User ID: {r['dummyUserId']}")
                else:
                    print(f"❌ {r['email']}: {r['error']}")
                    
        except Exception as e:
            print("\nFailed to parse API response:")
            print("Response:", response.text)
            raise e
                
    except Exception as e:
        print(f"\nError importing members: {str(e)}")
        exit(1)

if __name__ == "__main__":
    import_members() 

File: scripts/import-members.ts
import { read, utils, type WorkBook, type WorkSheet } from "xlsx";
import { createTRPCProxyClient, httpBatchLink } from "@trpc/client";
import { type AppRouter } from "@/server/api/root";
import superjson from "superjson";
import { z } from "zod";

// Define the expected Excel row structure
const ExcelRowSchema = z.object({
  Salutation: z.string().optional(),
  "English Name": z.string(),
  "Chinese Name": z.string().optional(),
  Email: z.string(),
  "Phone Number": z.union([z.string(), z.number()]).optional(),
  Class: z.union([z.string(), z.number()]),
  Faculty: z.string(),
  Major: z.string(),
  Industry: z.string().optional(),
  Employer: z.string().optional(),
  Position: z.string().optional(),
  "Membership Type": z
    .enum(["STUDENT", "ORDINARY_II", "ORDINARY_I", "HONORARY"])
    .optional(),
});

type ExcelRow = z.infer<typeof ExcelRowSchema>;

// Create a tRPC client
const trpc = createTRPCProxyClient<AppRouter>({
  links: [
    httpBatchLink({
      url: "http://localhost:3000/api/trpc",
      transformer: superjson,
    }),
  ],
});

async function importMembers() {
  try {
    console.log("Reading Excel file...");
    // Read the Excel file
    const workbook: WorkBook = read(
      "instructions/membersdata/Membership Book.xlsx",
    );
    const worksheet: WorkSheet = workbook.Sheets[workbook.SheetNames[0]!]!;
    const rawData = utils.sheet_to_json(worksheet);

    // Take only first 2 rows for testing
    const testData = rawData.slice(0, 2);
    console.log("\nProcessing first 2 rows:");
    console.log(JSON.stringify(testData, null, 2));

    // Validate and transform the data
    const validRows: ExcelRow[] = [];
    const invalidRows: { row: number; errors: string[] }[] = [];

    testData.forEach((row, index) => {
      console.log(`\nValidating row ${index + 1}:`);
      const result = ExcelRowSchema.safeParse(row);
      if (result.success) {
        console.log("✅ Row is valid");
        validRows.push(result.data);
      } else {
        console.log("❌ Row has validation errors");
        invalidRows.push({
          row: index + 2, // +2 because Excel is 1-based and we have a header row
          errors: result.error.errors.map(
            (e) => `${e.path.join(".")}: ${e.message}`,
          ),
        });
      }
    });

    // Log validation errors if any
    if (invalidRows.length > 0) {
      console.log("\nValidation errors found:");
      invalidRows.forEach(({ row, errors }) => {
        console.log(`Row ${row}:`);
        errors.forEach((error) => console.log(`  - ${error}`));
      });
      if (validRows.length === 0) {
        throw new Error("No valid rows found in Excel file");
      }
      console.log("\nProceeding with valid rows only...\n");
    }

    // Transform valid rows to match our API schema
    console.log("\nTransforming valid rows to API format:");
    const members = validRows.map((row) => {
      const member = {
        salutation: row.Salutation ?? "Mr",
        englishName: row["English Name"],
        chineseName: row["Chinese Name"],
        email: row.Email,
        phoneNumber: row["Phone Number"]?.toString() ?? "",
        class:
          typeof row.Class === "string" ? parseInt(row.Class, 10) : row.Class,
        faculty: row.Faculty,
        major: row.Major,
        industry: row.Industry,
        employer: row.Employer,
        position: row.Position,
        membershipType: row["Membership Type"] ?? "ORDINARY_I",
      };
      console.log(JSON.stringify(member, null, 2));
      return member;
    });

    console.log("\nImporting members to database...");
    // Import members using our API
    const result = await trpc.member.importMembers.mutate(members);

    console.log("\nImport completed:");
    console.log(`Total processed: ${result.totalProcessed}`);
    console.log(`Successful: ${result.successful}`);
    console.log(`Failed: ${result.failed}`);

    // Log detailed results
    console.log("\nDetailed results:");
    result.results.forEach((r) => {
      if (r.success) {
        console.log(`✅ ${r.email}: Successfully imported`);
        console.log(`  Member ID: ${r.memberId}`);
        console.log(`  Dummy User ID: ${r.dummyUserId}`);
      } else {
        console.log(`❌ ${r.email}: ${r.error}`);
      }
    });
  } catch (error) {
    if (error instanceof Error) {
      console.error("\nError importing members:", error.message);
    } else {
      console.error("\nUnknown error occurred during import");
    }
    process.exit(1);
  }
}

// Run the import
void importMembers();
